### PR TITLE
Adds snapshot API for MemPools to get pool memory segments

### PR DESCRIPTION
**Pull request by syed-ahmed**
Canonically, the snapshot API returns the entire memory state of the CUDACachingAllocator (using `get_all_blocks`). There is no API that can only return the memory state of a given pool. 

In this PR, we extend the functionality of snapshot API such that it can only return the memory addresses of an active pool. When snapshot API is called under a MemPoolContext, we only return the blocks that correspond to the pool id of the active pool.

Part of https://github.com/pytorch/pytorch/issues/124807.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133601

### CODE REVIEW

File: torch/cuda/memory.py [hunk 1]
```diff
@@ -980,6 +980,25 @@ def _get_current_allocator() -> _CUDAAllocator:
     return _CUDAAllocator(torch._C._cuda_getAllocator())
 
 
+class MemPoolContext(_MemPoolContext):
```
Comment (syed-ahmed) [inline torch/cuda/memory.py:983]: Needed to move this up, since snapshot below uses it.

**Comment by syed-ahmed** [comment]
@ezyang this is ready for review.

File: torch/cuda/memory.py [hunk 2]
```diff
@@ -1010,24 +1029,21 @@ def use_count(self) -> int:
...
-    def active_pool() -> Optional[_MemPool]:
-        r"""Returns the active MemPool"""
-        return _MemPoolContext.active_pool()
+        .. note::
+            See :ref:`cuda-memory-management` for more details about GPU memory
+            management.
+        """
+        ctx = MemPoolContext(self)
+        snapshot = torch.cuda.memory_snapshot()
+        del ctx
```
Comment (ezyang) [inline torch/cuda/memory.py:1045]: try..finally better in case memory_snapshot throws and something in backtrace keeps locals live

**Review by ezyang — APPROVED** [review]

File: torch/cuda/memory.py (hunk 2 above)
Comment (syed-ahmed) [reply torch/cuda/memory.py]: Added the try-finally. Merging now.

**Comment by syed-ahmed** [comment]
Thanks @wdvr for the logs! I've changed the `auto mempool_id` to just `mempool_id`. Hopefully that fixes the internal tests. Will try merging again once CI is green.

**Comment by syed-ahmed** [comment]
Merging, as the latest change should fix the internal shadow variable failures. Please feel free to revert again.

**Pull request closed without merging**